File: server/app/services/group_pagination.py
"""Version-2 collapse and presentation-only continuation rows for fixed row pages."""
import unicodedata

from ..models.domain import DomainError


def collapsed_group_keys(value, definition_version=1):
    if not isinstance(value, list) or len(value) > 1000 or any(not isinstance(key, str) or len(key) > 4096 or not key.startswith(('number:', 'string:', 'boolean:', 'null:', 'missing:')) for key in value):
        raise DomainError('invalid_group', 'Collapsed groups require at most 1000 bounded typed group keys.')
    if definition_version != 2 and value:
        raise DomainError('invalid_group', 'Group collapse requires definition version 2.')
    keys = [unicodedata.normalize('NFC', key) for key in value]
    if len(set(keys)) != len(keys):
        raise DomainError('invalid_group', 'Collapsed group keys must be unique.')
    return set(keys)


def paginate_group_rows(layout, capacity):
    groups = {row['row']: row for row in layout['rows']}
    counts = {'logicalGroupTotal': len(groups),
              'collapsedGroupTotal': sum(bool(row.get('collapsed')) for row in layout['rows']),
              'hiddenItemTotal': sum(row['recordCount'] for row in layout['rows'] if row.get('collapsed'))}
    if not groups:
        return {**layout, **counts}
    if capacity < 2 and any(not row.get('collapsed') for row in layout['rows']):
        raise DomainError('row_height_limit', 'A grouped timeline needs space for a group header and at least one readable record row.')
    positions, rows, offset, current = {}, [], 0, None
    # Materialize header slots once before layout publication, never copies of records.
    for row in range(layout['totalRows']):
        group = groups.get(row)
        if group is not None:
            if not group.get('collapsed') and offset % capacity == capacity - 1:
                offset += 1
            current = group
            rows.append({**group, 'row': offset, 'continuation': False})
        elif current is not None and offset % capacity == 0:
            rows.append({**current, 'row': offset, 'continuation': True})
            offset += 1
        positions[row] = offset
        offset += 1
    for item in layout['items']:
        item['row'] = positions[item['row']]
    for enclosure in layout.get('enclosures', []):
        enclosure['startRow'] = positions[enclosure['startRow']]
        enclosure['endRow'] = positions[enclosure['endRow'] - 1] + 1
    return {**layout, **counts, 'rows': rows, 'totalRows': offset}

File: client/src/timeline/group-pagination.js
const fail = (code, message) => { throw Object.assign(new Error(message), { code, status: 422 }); };

export function collapsedGroupKeys(value = [], definitionVersion = 1) {
  if (!Array.isArray(value) || value.length > 1000 || value.some(key => typeof key !== 'string' || Array.from(key).length > 4096 || !['number:', 'string:', 'boolean:', 'null:', 'missing:'].some(prefix => key.startsWith(prefix)))) fail('invalid_group', 'Collapsed groups require at most 1000 bounded typed group keys');
  if (definitionVersion !== 2 && value.length) fail('invalid_group', 'Group collapse requires definition version 2');
  const keys = value.map(key => key.normalize('NFC'));
  if (new Set(keys).size !== keys.length) fail('invalid_group', 'Collapsed group keys must be unique');
  return new Set(keys);
}

export function paginateGroupRows(layout, capacity) {
  const groups = new Map(layout.rows.map(row => [row.row, row]));
  const counts = {
    logicalGroupTotal: layout.rows.length,
    collapsedGroupTotal: layout.rows.filter(row => row.collapsed).length,
    hiddenItemTotal: layout.rows.reduce((total, row) => total + (row.collapsed ? row.recordCount : 0), 0),
  };
  if (!groups.size) return { ...layout, ...counts };
  if (capacity < 2 && layout.rows.some(row => !row.collapsed)) fail('row_height_limit', 'A grouped timeline needs space for a group header and at least one readable record row');
  const positions = [], rows = [];
  let offset = 0, current;
  // Insert presentation slots once, before pinning the layout; record IDs never repeat.
  for (let row = 0; row < layout.totalRows; row++) {
    const group = groups.get(row);
    if (group) {
      if (!group.collapsed && offset % capacity === capacity - 1) offset++;
      current = group;
      rows.push({ ...group, row: offset, continuation: false });
    } else if (current && offset % capacity === 0) {
      rows.push({ ...current, row: offset++, continuation: true });
    }
    positions[row] = offset++;
  }
  for (const item of layout.items) item.row = positions[item.row];
  for (const enclosure of layout.enclosures ?? []) {
    enclosure.startRow = positions[enclosure.startRow];
    enclosure.endRow = positions[enclosure.endRow - 1] + 1;
  }
  return { ...layout, ...counts, rows, totalRows: offset };
}
